Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp

```diff
@@ -0,0 +1,148 @@
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(MediaSourceHandle);
+
+struct MediaSourceHandle::SharedPrivate final : ThreadSafeRefCounted<SharedPrivate> {
+    static Ref<SharedPrivate> create(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher) { return adoptRef(*new SharedPrivate(mediaSource, WTFMove(dispatcher))); }
+
+    void dispatch(MediaSourceHandle::TaskType&& task, bool forceRunInWorker = false) const
+    {
+        m_dispatcher(WTFMove(task), forceRunInWorker);
+    }
+
+    void setMediaSourcePrivate(MediaSourcePrivate& mediaSourcePrivate)
+    {
+        Locker locker { m_lock };
+        m_private = mediaSourcePrivate;
+    }
+
+    RefPtr<MediaSourcePrivate> mediaSourcePrivate() const
+    {
+        Locker locker { m_lock };
+        return m_private.get();
+    }
+
+    friend class MediaSourceHandle;
+    SharedPrivate(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher)
+        : m_client(mediaSource.client().get())
+        , m_isManaged(mediaSource.isManaged())
+        , m_dispatcher(WTFMove(dispatcher))
+    {
+    }
+    mutable Lock m_lock;
+    ThreadSafeWeakPtr<MediaSourcePrivate> m_private WTF_GUARDED_BY_LOCK(m_lock);
+    std::atomic<bool> m_hasEverBeenAssignedAsSrcObject { false };
+    ThreadSafeWeakPtr<MediaSourcePrivateClient> m_client;
+    const bool m_isManaged;
+    const MediaSourceHandle::DispatcherType m_dispatcher;
+};
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher)
+{
+    return adoptRef(*new MediaSourceHandle(mediaSource, WTFMove(dispatcher)));
+}
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(Ref<MediaSourceHandle>&& other)
+{
+    other->setDetached(false);
+    return other;
+}
+
+MediaSourceHandle::MediaSourceHandle(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher)
+    : m_private(MediaSourceHandle::SharedPrivate::create(mediaSource, WTFMove(dispatcher)))
+{
+}
+
+MediaSourceHandle::MediaSourceHandle(MediaSourceHandle& other)
+    : m_private(other.m_private)
+{
+    ASSERT(!other.m_detached);
+    other.m_detached = true;
+}
+
+MediaSourceHandle::~MediaSourceHandle() = default;
+
+bool MediaSourceHandle::canDetach() const
+{
+    return !isDetached() && !m_private->m_hasEverBeenAssignedAsSrcObject;
+}
+
+void MediaSourceHandle::setHasEverBeenAssignedAsSrcObject()
+{
+    m_private->m_hasEverBeenAssignedAsSrcObject = true;
+}
+
+bool MediaSourceHandle::hasEverBeenAssignedAsSrcObject() const
+{
+    return m_private->m_hasEverBeenAssignedAsSrcObject;
+}
+
+bool MediaSourceHandle::isManaged() const
+{
+    return m_private->m_isManaged;
+}
+
+void MediaSourceHandle::ensureOnDispatcher(MediaSourceHandle::TaskType&& task) const
+{
+    m_private->dispatch(WTFMove(task));
+}
+
+Ref<DetachedMediaSourceHandle> MediaSourceHandle::detach()
+{
+    return adoptRef(*new MediaSourceHandle(*this));
```

Comment:
It would read better if we would do something like:
```
m_detached = true;
return ...
```
See above where it seems we could just send a ref to the private.